src/runtime_src/core/pcie/linux/system_linux.cpp

```diff
@@ -98,26 +110,76 @@ static std::string machine_info()
   return model;
 }
 
-static std::vector<std::weak_ptr<xrt_core::device_linux>> mgmtpf_devices(16); // fix size
-static std::vector<std::weak_ptr<xrt_core::device_linux>> userpf_devices(16); // fix size
-static std::map<xrt_core::device::handle_type, std::weak_ptr<xrt_core::device_linux>> userpf_device_map;
-
 }
 
 namespace xrt_core {
 
+void
+system_linux::
+register_driver(std::shared_ptr<pcidrv::pci_driver> driver)
+{
+  driver_list.push_back(driver);
+  if (driver->is_user())
+    driver->scan_devices(user_ready_list, user_nonready_list);
+  else
+    driver->scan_devices(mgmt_ready_list, mgmt_nonready_list);
+}
+
+std::shared_ptr<pcidev::pci_device>
+system_linux::
+get_pcidev(unsigned index, bool is_user) const
+{
+  if (is_user) {
+    if (index < user_ready_list.size())
+      return user_ready_list[index];
+    return user_nonready_list[index - user_ready_list.size()];
+  }
+
+  if (index < mgmt_ready_list.size())
+    return mgmt_ready_list[index];
+  return mgmt_nonready_list[index - mgmt_ready_list.size()];
+}
+
+size_t
+system_linux::
+get_num_dev_ready(bool is_user) const
+{
+  if (is_user)
+    return user_ready_list.size();
+  return mgmt_ready_list.size();
+}
+
+size_t
+system_linux::
+get_num_dev_total(bool is_user) const
+{
+  if (is_user)
+    return user_ready_list.size() + user_nonready_list.size();
+  return mgmt_ready_list.size() + mgmt_nonready_list.size();
+}
+
+system_linux::
+system_linux()
+{
+  if (singleton)
+    throw std::runtime_error("More than one SHIM registered!");
+  singleton = this;
```

Comment:
Argh. Avoid teleportation.
Move this logic to the singleton handler instead as Kate Gregory says.

Reply:
@keryell Yeah, I think that logic is there already.  It is strictly enforced that only one singleton exists and this is handled at base class level.  One problem we are catching is when users link the "xcl" application against xrt_core and then runs hw or sw emulation which tries to load libxrt_hwemu or swemu at runtime; this would imply two singletons, which is currently not supported because two core libraries cannot be loaded at the same time as long as we support xcl APIs.